services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py

```diff
@@ -73,6 +74,70 @@ async def get_comp_tasks(
 
         return tasks
 
+    # pylint: disable=unused-argument
+    @run_sequentially_in_context(target_args=["str_project_uuid", "node_id"])
+    async def _upsert_node(
+        self,
+        project: ProjectAtDB,
+        director_client: DirectorV0Client,
+        publish: bool,
+        node_id: str,
+        internal_id: int,
+        str_project_uuid: str,
+    ) -> None:
+        """Upsert data for each node sequentially"""
+        node: Node = project.workbench[node_id]
+
+        service_key_version = ServiceKeyVersion(
+            key=node.key,
+            version=node.version,
+        )
+        node_class = to_node_class(service_key_version.key)
+        node_details: ServiceDockerData = None
+        node_extras: ServiceExtras = None
+        if node_class == NodeClass.FRONTEND:
+            node_details = _get_fake_service_details(service_key_version)
+        else:
+            node_details = await director_client.get_service_details(
+                service_key_version
+            )
+            node_extras: ServiceExtras = await director_client.get_service_extras(
+                service_key_version
+            )
+        requires_mpi = False
+        requires_gpu = False
+        if node_extras:
+            requires_gpu = node_extras.node_requirements == NodeRequirement.GPU
+            requires_mpi = node_extras.node_requirements == NodeRequirement.MPI
+        image = Image(
+            name=service_key_version.key,
+            tag=service_key_version.version,
+            requires_gpu=requires_gpu,
+            requires_mpi=requires_mpi,
+        )
+
+        comp_state = RunningState.PUBLISHED if publish else RunningState.NOT_STARTED
+        task_db = CompTaskAtDB(
+            project_id=project.uuid,
+            node_id=node_id,
+            schema=NodeSchema(inputs=node_details.inputs, outputs=node_details.outputs),
+            inputs=node.inputs,
+            outputs=node.outputs,
+            image=image,
+            submit=datetime.utcnow(),
+            state=(
+                comp_state
+                if node_class == NodeClass.COMPUTATIONAL
+                else RunningState.NOT_STARTED
+            ),
+            internal_id=internal_id,
+            node_class=node_class,
+        )
+
+        await self.connection.execute(
+            insert(comp_tasks).values(**task_db.dict(by_alias=True, exclude_unset=True))
+        )
+
     @log_decorator(logger=logger)
     async def upsert_tasks_from_project(
```

Comment:
hmm I think the sequential decorator should be on that one don't you think?
what happens if we delete the tasks while the other is inserting?

Reply:
I overlooked the delete. At this point the best thing to do is to lock the entire `upsert_tasks_from_project` based on `project_id`. Will guarantee that the delete recreate happens without external interventions.

Reply:
👍